reana_client/validation/complexity.py

```diff
@@ -0,0 +1,211 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of REANA.
+# Copyright (C) 2021 CERN.
+#
+# REANA is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""REANA workflow complexity estimation."""
+
+import os
+
+from reana_client.printer import display_message
+
+
+def estimate_complexity(workflow_type, reana_yaml):
+    """Estimate complexity in REANA workflow.
+
+    :param workflow_type: A supported workflow specification type.
+    :param reana_yaml: REANA YAML specification.
+    """
+
+    def build_estimator(workflow_type, reana_yaml):
+        if workflow_type == "serial":
+            return SerialComplexityEstimator(reana_yaml)
+        if workflow_type == "yadage":
+            return YadageComplexityEstimator(reana_yaml)
+        if workflow_type == "cwl":
+            return CWLComplexityEstimator(reana_yaml)
+
+    estimator = build_estimator(workflow_type, reana_yaml)
+    complexity = estimator.estimate_complexity()
+
+    display_message(
+        "Estimated workflow complexity: \n{0}".format(complexity),
+        msg_type="info",
+        indented=True,
+    )
+
+
+class ComplexityEstimatorBase:
+    """REANA workflow complexity estimator base class."""
+
+    def __init__(self, reana_yaml, initial_step="init"):
```

Comment:
Optional: I wonder whether `initial_step` should be a parameter of the initaliser, or only of estimate_complexity method?  If the former, one has to initialise class twice for each run.  If the latter, one could initialise the class and launch several estimations for various starting steps. This could be useful for restarting workflow from given later step, for example.

Note: Since we are estimating only the workflow launching complexity, we are not probable to call this function several times... so it is fully OK for the code to go "as is".